packages/storage/src/append_store.rs

```diff
@@ -3,378 +3,256 @@
-            Self::new(storage, &len_vec)
+impl<'a, T: Serialize + DeserializeOwned, Ser: Serde> AppendStore<'a, T, Ser> {
+    /// gets the length from storage, and otherwise sets it to 0
+    pub fn get_len<S: ReadonlyStorage>(&self, storage: &S) -> StdResult<u32> {
+        let len_key = [self.as_slice(), LEN_KEY].concat();
+        if let Some(len_vec) = storage.get(&len_key) {
+            let len_bytes = len_vec.as_slice().try_into().map_err(|err| StdError::parse_err("u32", err))?;
+            let len = u32::from_be_bytes(len_bytes);
+            Ok(len)
         } else {
-            let len_vec = 0_u32.to_be_bytes();
-            storage.set(LEN_KEY, &len_vec);
-            Self::new(storage, &len_vec)
+            Ok(0)
         }
     }
-
-    /// Try to use the provided storage as an AppendStore.
-    /// This method allows choosing the serialization format you want to use.
-    ///
-    /// Returns None if the provided storage doesn't seem like an AppendStore.
-    /// Returns Err if the contents of the storage can not be parsed.
-    pub fn attach_with_serialization(storage: &'a mut S, _ser: Ser) -> Option<StdResult<Self>> {
-        let len_vec = storage.get(LEN_KEY)?;
-        Some(Self::new(storage, &len_vec))
-    }
-
-    fn new(storage: &'a mut S, len_vec: &[u8]) -> StdResult<Self> {
-        let len_array = len_vec
-            .try_into()
-            .map_err(|err| StdError::parse_err("u32", err))?;
-        let len = u32::from_be_bytes(len_array);
-
-        Ok(Self {
-            storage,
-            item_type: PhantomData,
-            serialization_type: PhantomData,
-            len,
-        })
-    }
-
-    pub fn len(&self) -> u32 {
-        self.len
-    }
-
-    pub fn is_empty(&self) -> bool {
-        self.len == 0
-    }
-
-    pub fn storage(&mut self) -> &mut S {
-        self.storage
-    }
-
-    pub fn readonly_storage(&self) -> &S {
-        self.storage
-    }
-
-    /// Return an iterator over the items in the collection
-    pub fn iter(&self) -> Iter<T, S, Ser> {
-        self.as_readonly().iter()
-    }
-
-    /// Get the value stored at a given position.
-    ///
-    /// # Errors
-    /// Will return an error if pos is out of bounds or if an item is not found.
-    pub fn get_at(&self, pos: u32) -> StdResult<T> {
-        self.as_readonly().get_at(pos)
+    /// checks if the collection has any elements
+    pub fn is_empty<S: ReadonlyStorage>(&self, storage: &S) -> StdResult<bool> {
+        Ok(self.get_len(storage)? == 0)
     }
-
-    fn get_at_unchecked(&self, pos: u32) -> StdResult<T> {
-        self.as_readonly().get_at_unchecked(pos)
+    /// gets the element at pos if within bounds
+    pub fn get_at<S: ReadonlyStorage>(&self, storage: &S, pos: u32) -> StdResult<T> {
+        let len = self.get_len(storage)?;
+        if pos > len { return Err(StdError::generic_err("AppendStore access out of bounds")); }
+        self.get_at_unchecked(storage, pos)
     }
```

Comment:
This kind of thing makes fetching less efficient than before... but i guess there's no way to cache the length without mutating `self`, and we don't want to do that because we want instances of AppendStore to be const... I guess that's fine for now

Reply:
On second thought, I guess we can still cache the length inside the store if we're ok with it being a static instead of a const 🤔 it doesn't even have to be a mutable static if the length field is `len: Cell<u32>`

Reply:
I added caching with Cell as suggested in the latest push